File: symmetric_tree/KHH/SymmetricTree.java
import java.util.*;
import java.io.*;

class Solution
{
    public boolean isSymmetric(TreeNode root) {
        if(null == root) return true;
        else return isEquals(root.left,  root.right);
    }
    public static boolean isEquals(TreeNode left, TreeNode right)
    {
        if (null == left)
        {
            if (null == right) return true;
            else return false;
        }
        else
        {
            if( null == right) 
                return false;
            else if(left.val == right.val)
            {
                return isEquals(left.left, right.right) && isEquals(left.right, right.left);
            }
            else return false;
        }
    }
}


/* 
* Main 
*/
public class SymmetricTree {
    public static TreeNode stringToTreeNode(String input) {
        input = input.trim();
        input = input.substring(1, input.length() - 1);
        if (input.length() == 0) {
            return null;
        }
    
        String[] parts = input.split(",");
        String item = parts[0];
        TreeNode root = new TreeNode(Integer.parseInt(item));
        Queue<TreeNode> nodeQueue = new LinkedList<>();
        nodeQueue.add(root);
    
        int index = 1;
        while(!nodeQueue.isEmpty()) {
            TreeNode node = nodeQueue.remove();
    
            if (index == parts.length) {
                break;
            }
    
            item = parts[index++];
            item = item.trim();
            if (!item.equals("null")) {
                int leftNumber = Integer.parseInt(item);
                node.left = new TreeNode(leftNumber);
                nodeQueue.add(node.left);
            }
    
            if (index == parts.length) {
                break;
            }
    
            item = parts[index++];
            item = item.trim();
            if (!item.equals("null")) {
                int rightNumber = Integer.parseInt(item);
                node.right = new TreeNode(rightNumber);
                nodeQueue.add(node.right);
            }
        }
        return root;
    }
    
    public static String booleanToString(boolean input) {
        return input ? "True" : "False";
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        String line;
        while ((line = in.readLine()) != null) {
            TreeNode root = stringToTreeNode(line);
            
            boolean ret = new Solution().isSymmetric(root);
            
            String out = booleanToString(ret);
            
            System.out.print(out);
        }
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

File: symmetric_tree/KTH/SymmetricTree.kt
import java.util.*
import kotlin.math.log2

fun main(args:Array<String>){
    val root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(2)
    root.left?.let {
       it.left = TreeNode(3)
        it.right = TreeNode(4)
    }
    root.right?.let {
        it.left = TreeNode(4)
        it.right = TreeNode(3)
    }
    Solution().isSymmetric(root)
}
  class TreeNode(var `val`: Int = 0) {
      var left: TreeNode? = null
     var right: TreeNode? = null
 }

class Solution {
    fun isSymmetric(root: TreeNode?): Boolean {
        if(root == null)
            return true
        val list = ArrayList<ArrayList<Int?>>()
        bfs(root, list)
        if(list.size == 1)
            return true
        for( i in list){
            for(j in 0 until i.size){
                if(i[j] != i[i.size -1 -j] )
                    return false
            }
        }
        return true
    }
    private fun bfs(root : TreeNode?, list:ArrayList<ArrayList<Int?>>)
    {
        root?.let {
            val q = ArrayList<TreeNode?>()
            var secQ:Queue<TreeNode?>
            q.add(it)
            while(q.any()){
                secQ = LinkedList(q)
                q.clear()
                list.add(ArrayList())
                while(secQ.any()){
                    val node = secQ.poll()
                    list.last().add(node?.`val`)
                    node?.let {
                        q.add(it.left)
                        q.add(it.right)
                    }
                }
            }
        }
    }
}
